refactor(linter): add `CompositeFix::merge_fixes_fallible` method

overlookmotel opened the pull request:
Add a method `CompositeFix::merge_fixes_fallible` which returns `Result::Err` in case of overlapping fixes or other illegal states, instead of panicking in debug mode.

`CompositeFix::merge_fixes` does the same as it did before, but uses `merge_fixes_fallible` internally.

Use an error type `MergeFixesError` instead of `String` to avoid adding string formatting overhead when the primary use case in release mode is to ignore the errors anyway.

`merge_fixes_fallible` will be used in JS plugins where we don't want to panic on these illegal states, but output a diagnostic instead.

overlookmotel commented:
* **#14093** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/14093?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/14093?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=14093).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

camc314 commented:
### Merge activity

* **Sep 24, 10:32 PM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel committed:
refactor(linter): add `CompositeFix::merge_fixes_fallible` method (#14093)

Add a method `CompositeFix::merge_fixes_fallible` which returns `Result::Err` in case of overlapping fixes or other illegal states, instead of panicking in debug mode.

`CompositeFix::merge_fixes` does the same as it did before, but uses `merge_fixes_fallible` internally.

Use an error type `MergeFixesError` instead of `String` to avoid adding string formatting overhead when the primary use case in release mode is to ignore the errors anyway.

`merge_fixes_fallible` will be used in JS plugins where we don't want to panic on these illegal states, but output a diagnostic instead.
